lib/gooddata/models/client.rb

```diff
@@ -33,32 +33,30 @@ def [](id, opts = {})
         domain = opts[:domain]
         segment = opts[:segment]
         fail ArgumentError, 'No :domain specified' if domain.nil?
-        fail ArgumentError, 'No :segment specified' if domain.nil?
-
         client = domain.client
         fail ArgumentError, 'No client specified' if client.nil?
+        data_product = opts[:data_product] || segment.data_product
 
+        uri = base_uri(domain, data_product)
         if id == :all
-          data_product = segment.data_product
-          tenants_uri = GoodData::DataProduct::ONE_DATA_PRODUCT_PATH % { domain_name: domain.name, id: data_product.data_product_id }
-          tenants_uri += "/clients?segment=#{CGI.escape(segment.segment_id)}"
+          uri += "?segment=#{CGI.escape(segment.segment_id)}" if segment
           Enumerator.new do |y|
             loop do
-              res = client.get tenants_uri
+              res = client.get uri
```

Comment:
Refers to 'client' more than self (maybe move it to another class?) (FeatureEnvy)